# Tarea 3.7

Ejercicio 1.3.1 del libro de Burden: Use números de punto flotante de 16 bits para calcular las siguientes sumas. Explique por qué ambos métodos en cada caso dan resultados diferentes y cuál es la más correcta.

* $\sum_{n = 1}^{100}\frac{1}{n^2}$ primero de la forma $1 + \frac{1}{4} + \frac{1}{9} + \dots + \frac{1}{10^4}$ y luego en la forma $\frac{1}{10^4} + \dots + \frac{1}{9} + \frac{1}{4} + 1$.

* $\sum_{n = 1}^{100}\frac{1}{n^3}$ primero de la forma $1 + \frac{1}{8} + \frac{1}{27} + \dots + \frac{1}{10^6}$ y luego en la forma $\frac{1}{10^6} + \dots + \frac{1}{27} + \frac{1}{8} + 1$.

## Solución  

### Primer caso

Primero haremos la sumatoria con los términos ordenados del 1 al 100: $\sum_{n = 1}^{100}\frac{1}{n^2}$ con la forma $1 + \frac{1}{4} + \frac{1}{9} + \dots + \frac{1}{10^4}$

In [1]:
import numpy as np

In [2]:
#Generamos un ciclo for para hacer la sumatoria con números de 16 bits.
suma1 =np.float16(0)
for i in range(100):
    i=np.float16(i)+np.float16(1)
    suma1+=np.float16(1)/(i**np.float16(2))

In [3]:
#Obtenemos:
suma1

1.627

Ahora hacemos la misma sumatoria, pero partiendo del último termino hacia el primero:   $\sum_{n = 1}^{100}\frac{1}{n^2}$de la forma $\frac{1}{10^4} + \dots + \frac{1}{9} + \frac{1}{4} + 1$

In [4]:
#Nuevamente generamos un ciclo for para hacer la sumatoria con números de 16 bits pero invirtiendo el orden  de la suma.
#Con el comando "reversed" hacemos que el ciclo for genere los números desde el 100 hasta el 1.
suma2 =np.float16(0)
for j in reversed(range(100)):
    j=np.float16(j)+np.float16(1)
    suma2+=np.float16(1)/(j**np.float16(2))

In [5]:
#Obtenemos:
suma2

1.635

 El valor real de la sumatoria es aproximadamente de 1.63498, por lo cual el segundo método es más preciso al darnos un valor  de 1.635


### Segundo caso


* $\sum_{n = 1}^{100}\frac{1}{n^3}$ primero de la forma $1 + \frac{1}{8} + \frac{1}{27} + \dots + \frac{1}{10^6}$ 

In [6]:
#Repetimos el mismo procedimiento que en la "suma1", solo que cambiamos el exponente de la sumatoria. 
suma3 =np.float16(0)
for i in range(100):
    i=np.float16(i)+np.float16(1)
    suma3+=np.float16(1)/(i**np.float16(3))

<ipython-input-6-bcfefb522bc9>:5: RuntimeWarning: overflow encountered in half_scalars
  suma3+=np.float16(1)/(i**np.float16(3))


In [7]:
#Obtenemos:
suma3

1.199

* $\sum_{n = 1}^{100}\frac{1}{n^3}$ ahora con la forma $\frac{1}{10^6} + \dots + \frac{1}{27} + \frac{1}{8} + 1$

In [8]:
#Repetimos el mismo procedimiento que en la "suma2", solo que cambiamos el exponente de la sumatoria.
suma4 =np.float16(0)
for j in reversed(range(100)):
    j=np.float16(j)+np.float16(1)
    suma4+=np.float16(1)/(j**np.float16(3))

<ipython-input-8-e29af3cdd604>:5: RuntimeWarning: overflow encountered in half_scalars
  suma4+=np.float16(1)/(j**np.float16(3))


In [9]:
#Obtenemos:
suma4

1.202

Nuevamente el segundo método para hacer la sumatoria es más preciso, ya que aproximadamente el valor real de la sumatoria es de 1.20200

### Explicacion

Debe existir un error de redondeo si en cada caso ambos métodos dan resultados diferentes, al parecer resulta más correcto el segundo método (donde hacemos la sumatoria desde el ultimo termino al primero) y esto debido a que estamos sumando los términos más pequeños al inicio de la sumatoria, resultando en un error de redondeo menor al que obtenemos al hacer la sumatoria de manera tradicional.

# Calificación: 7.0

Faltó solo ser más preciso: En el primer caso, en los últimos términos de la sumatoria, se suman números muy diferentes (la suma va siendo un número grande y el siguiente número es pequeño). En estas situaciones se incurre siempre en un mayor error de redondeo (el computador al sumar primero convierte ambos números al exponente del número más grande).